pkg/controllers/config/configsync.go

```diff
@@ -0,0 +1,220 @@
+	"k8s.io/klog/v2"
+)
+
+// This ConfigSyncController translates the ConfigMap provided by the user
+// containing configuration information for the Cinder CSI driver.
+type ConfigSyncController struct {
+	operatorClient       v1helpers.OperatorClient
+	kubeClient           kubernetes.Interface
+	configMapLister      corelisters.ConfigMapLister
+	infrastructureLister configv1listers.InfrastructureLister
+	eventRecorder        events.Recorder
+}
+
+const (
+	sourceConfigKey = "config"
+	targetConfigKey = "cloud.conf"
+
+	infrastructureResourceName = "cluster"
+)
+
+func NewConfigSyncController(
+	operatorClient v1helpers.OperatorClient,
+	kubeClient kubernetes.Interface,
+	informers v1helpers.KubeInformersForNamespaces,
+	configInformers configinformers.SharedInformerFactory,
+	resyncInterval time.Duration,
+	eventRecorder events.Recorder) factory.Controller {
+
+	// Read configmap from user-managed namespace and save the translated one
+	// to the operator namespace
+	configMapInformer := informers.InformersFor(util.OpenShiftConfigNamespace)
+	c := &ConfigSyncController{
+		operatorClient:       operatorClient,
+		kubeClient:           kubeClient,
+		configMapLister:      configMapInformer.Core().V1().ConfigMaps().Lister(),
+		infrastructureLister: configInformers.Config().V1().Infrastructures().Lister(),
+		eventRecorder:        eventRecorder.WithComponentSuffix("ConfigSync"),
+	}
+	return factory.New().WithSync(c.sync).ResyncEvery(resyncInterval).WithSyncDegradedOnError(operatorClient).WithInformers(
+		operatorClient.Informer(),
+		configMapInformer.Core().V1().ConfigMaps().Informer(),
+	).ToController("ConfigSync", eventRecorder)
+}
+
+func (c *ConfigSyncController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	opSpec, _, _, err := c.operatorClient.GetOperatorState()
+	if err != nil {
+		return err
+	}
+	if opSpec.ManagementState != operatorv1.Managed {
+		return nil
+	}
+
+	isMultiAZDeployment, err := isMultiAZDeployment()
+	if err != nil {
+		return err
+	}
+
+	infra, err := c.infrastructureLister.Get(infrastructureResourceName)
+	if err != nil {
+		return err
+	}
+
+	sourceConfig, err := c.configMapLister.ConfigMaps(util.OpenShiftConfigNamespace).Get(infra.Spec.CloudConfig.Name)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// TODO: report error after some while?
+			klog.V(2).Infof("Waiting for config map %s from %s", infra.Spec.CloudConfig.Name, util.OpenShiftConfigNamespace)
+			return nil
+		}
+		return err
+	}
+
+	targetConfig, err := translateConfigMap(sourceConfig, isMultiAZDeployment)
+	if err != nil {
+		return err
+	}
+
+	_, _, err = resourceapply.ApplyConfigMap(ctx, c.kubeClient.CoreV1(), c.eventRecorder, targetConfig)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func isMultiAZDeployment() (bool, error) {
+	ci, err := getCloudInfo()
+	if err != nil {
+		return false, fmt.Errorf("couldn't collect info about cloud availability zones: %w", err)
+	}
+	klog.V(2).Infof("found cloud info: %+v", ci)
```

Comment:
We'll have to remove this debug logs before merging.

Reply:
I've kept them for now while I try to debug the frequent calls but I have added a `TODO`.